src/axom/sidre/tests/sidre_read_write_userdefined_data.cpp

```diff
@@ -0,0 +1,706 @@
+}
+
+//------------------------------------------------------------------------------
+
+template <typename T, int DIM>
+void test_user_defined_data()
+{
+  // populate data
+  constexpr IndexType size = 10;
+  axom::Array<T, DIM> states(size, size);
+  fill_array(states, 0, true);
+
+  // Create datastore
+  DataStore ds;
+  Group* root = ds.getRoot();
+
+  // get size
+  auto num_states = static_cast<IndexType>(states.size());
+  auto state_size = static_cast<IndexType>(sizeof(*(states.begin())));
+  auto total_size = num_states * state_size;
+
+  SLIC_INFO(axom::fmt::format("Num of States={}", num_states));
+  SLIC_INFO(axom::fmt::format("State Size={}", state_size));
+  SLIC_INFO(axom::fmt::format("Total Size={}", total_size));
+  SLIC_INFO(
+    axom::fmt::format("Total Size/State Size={}", total_size / state_size));
+
+  // write shape
+  root->createViewScalar("num_states", num_states);
+  root->createViewScalar("state_size", state_size);
+  root->createViewScalar("total_size", total_size);
+
+  // write data to datastore as bytes
+  View* states_view =
+    root->createViewAndAllocate("states", axom::sidre::UINT8_ID, total_size);
+  std::uint8_t* sidre_state_data = states_view->getData();
+  memcpy(sidre_state_data, states.data(), static_cast<std::size_t>(total_size));
+
+  // mess with data before overriding it back to original in sidre
+  fill_array(states, -1, false);
+
+  // Copy original data back over local data
+  memcpy(states.data(), sidre_state_data, static_cast<std::size_t>(total_size));
+
+  // Test data is back to original
+  check_array(states);
+}
+
+template <typename T, int DIM>
+void test_external_user_defined_data()
+{
+  // populate data
+  constexpr IndexType size = 10;
+  axom::Array<T, DIM> states(size, size);
+  fill_array(states, 0, true);
+
+  // Create datastore
+  DataStore ds;
+  Group* root = ds.getRoot();
+
+  // get size
+  auto num_states = static_cast<IndexType>(states.size());
+  auto state_size = static_cast<IndexType>(sizeof(*(states.begin())));
+  auto total_size = num_states * state_size;
+  auto num_uint8s = total_size / sizeof(std::uint8_t);
+
+  SLIC_INFO(axom::fmt::format("Num of States={}", num_states));
+  SLIC_INFO(axom::fmt::format("State Size={}", state_size));
+  SLIC_INFO(axom::fmt::format("Total Size={}", total_size));
+  SLIC_INFO(
+    axom::fmt::format("Total Size/State Size={}", total_size / state_size));
+  SLIC_INFO(axom::fmt::format("Num of uint8={}", num_uint8s));
+
+  // write shape
+  root->createViewScalar("num_states", num_states);
+  root->createViewScalar("state_size", state_size);
+  root->createViewScalar("total_size", total_size);
+
+  // Add states as an external buffer
+  View* states_view = root->createView("states");
+  states_view->setExternalDataPtr(axom::sidre::UINT8_ID,
+                                  num_uint8s,
+                                  states.data());
+
+  // Save the array data in to a file
```

Comment:
Minor: The following might be clearer:
```suggestion
  // Save the array data into a file
```